sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/BatchObserverSdk.java

```diff
@@ -0,0 +1,248 @@
+import io.opentelemetry.sdk.metrics.aggregator.NoopAggregator;
+import io.opentelemetry.sdk.metrics.data.MetricData;
+import io.opentelemetry.sdk.metrics.data.MetricData.Descriptor;
+import io.opentelemetry.sdk.metrics.data.MetricData.Point;
+import io.opentelemetry.sdk.resources.Resource;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.locks.ReentrantLock;
+
+/** Implementation of the {@link BatchObserver}. */
+final class BatchObserverSdk extends AbstractInstrument implements BatchObserver {
+
+  private final MeterSdk meter;
+  private final BatchObserverFunction function;
+  private final BatchObserverBatcher batcher;
+  private final ReentrantLock collectLock = new ReentrantLock();
+
+  BatchObserverSdk(
+      InstrumentDescriptor descriptor,
+      MeterProviderSharedState meterProviderSharedState,
+      MeterSharedState meterSharedState,
+      MeterSdk meterSdk,
+      BatchObserverBatcher batcher,
+      BatchObserverFunction function) {
+    super(descriptor, meterProviderSharedState, meterSharedState, new ActiveBatcher(batcher));
+    this.meter = meterSdk;
+    this.batcher = batcher;
+    this.function = function;
+  }
+
+  @Override
+  public DoubleSumObserver.Builder doubleSumObserverBuilder(String name) {
+    return this.meter.doubleSumObserverBuilder(name);
+  }
+
+  @Override
+  public LongSumObserver.Builder longSumObserverBuilder(String name) {
+    return this.meter.longSumObserverBuilder(name);
+  }
+
+  @Override
+  public DoubleUpDownSumObserver.Builder doubleUpDownSumObserverBuilder(String name) {
+    return this.meter.doubleUpDownSumObserverBuilder(name);
+  }
+
+  @Override
+  public LongUpDownSumObserver.Builder longUpDownSumObserverBuilder(String name) {
+    return this.meter.longUpDownSumObserverBuilder(name);
+  }
+
+  @Override
+  public DoubleValueObserver.Builder doubleValueObserverBuilder(String name) {
+    return this.meter.doubleValueObserverBuilder(name);
+  }
+
+  @Override
+  public LongValueObserver.Builder longValueObserverBuilder(String name) {
+    return this.meter.longValueObserverBuilder(name);
+  }
+
+  @Override
+  List<MetricData> collectAll() {
+    final List<MetricData> metricData = new ArrayList<>();
+    if (function == null) {
+      return Collections.emptyList();
+    }
+    collectLock.lock();
+    try {
+      function.observe(
+          (labels, observations) -> {
+            batcher.setLabels(labels);
+            for (Observation observation : observations) {
+              ObservationType type = observation.getType();
+              Aggregator aggregator;
+              Descriptor descriptor;
+              if (type == ObservationType.LONG_OBSERVATION) {
+                LongObservation longObservation = (LongObservation) observation;
+                aggregator = longObservation.getAggregator();
+                aggregator.recordLong(longObservation.getValue());
```

Comment:
rather than requiring a 2-step process here, could we have the LongObservation just have a method like `recordLong` that returns the Aggregator?